### PR TITLE
Delete project note endpoint and tests.

### PR DESCRIPTION
**Pull request by tim-wilde-nimble**

### CODE REVIEW

File: Dfe.Academies.Academisation.WebApi/Controllers/LegacyProjectController.cs [hunk 1]
```diff
@@ -141,5 +144,32 @@ public async Task<ActionResult> AddNote(int id, AddNoteRequest note)
 				_ => throw new NotImplementedException()
 			};
 		}
+
+		/// <summary>
+		///     Deletes the provided note from the project with the specified ID
```
Comment (jordan-beard) [inline Dfe.Academies.Academisation.WebApi/Controllers/LegacyProjectController.cs:149]: Small thing - this summary is not what is happening. It looks like you are deleting the note that matches all criteria rather than id.

**Review by dneed-nimble — APPROVED** [review]

File: Dfe.Academies.Academisation.Service/Commands/Legacy/Project/LegacyProjectDeleteNoteCommand.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
+﻿using Dfe.Academies.Academisation.Core;
+using Dfe.Academies.Academisation.Data;
+using Dfe.Academies.Academisation.Data.ProjectAggregate;
+using Dfe.Academies.Academisation.Domain.Core.ProjectAggregate;
+using Dfe.Academies.Academisation.IDomain.ProjectAggregate;
+using Dfe.Academies.Academisation.IService.Commands.Legacy.Project;
+using Dfe.Academies.Academisation.IService.ServiceModels.Legacy.ProjectAggregate;
+using Microsoft.EntityFrameworkCore;
+
+namespace Dfe.Academies.Academisation.Service.Commands.Legacy.Project
```
Comment (jordan-beard) [inline Dfe.Academies.Academisation.Service/Commands/Legacy/Project/LegacyProjectDeleteNoteCommand.cs:10]: I think the original intention was to have code in the service project call into the data project, which could use the AcademisationContext directly. Though it is a bit overcomplicated. This API needs a refactor to make it easier to use and consistent.

**Review by jordan-beard — APPROVED** [review]

File: Dfe.Academies.Academisation.WebApi/Controllers/LegacyProjectController.cs (hunk 1 above)
Comment (tim-wilde-nimble) [reply Dfe.Academies.Academisation.WebApi/Controllers/LegacyProjectController.cs:149]: "from the project with the specified ID" - the rest is DDD impl. (the value object doesn't have an ID)

File: Dfe.Academies.Academisation.Service/Commands/Legacy/Project/LegacyProjectDeleteNoteCommand.cs (hunk 2 above)
Comment (tim-wilde-nimble) [reply Dfe.Academies.Academisation.Service/Commands/Legacy/Project/LegacyProjectDeleteNoteCommand.cs:7]: Refactored to keep the AcademisationContext in the Data project.

File: Dfe.Academies.Academisation.Data/ProjectAggregate/ProjectNoteAddCommand.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+		private readonly AcademisationContext _context;
+
+		public ProjectNoteAddCommand(AcademisationContext context)
+		{
+			_context = context;
+		}
+
+		public async Task<CommandResult> Execute(int projectId, ProjectNote note)
+		{
+			if (_context.Projects.Any(x => x.Id == projectId) is false)
```
Comment (dneed-nimble) [inline Dfe.Academies.Academisation.Data/ProjectAggregate/ProjectNoteAddCommand.cs:18]: Why pass the Id separately and not on the ProjectNote?

File: Dfe.Academies.Academisation.IService/Commands/Legacy/Project/ILegacyProjectDeleteNoteCommand.cs [hunk 4]
```diff
@@ -0,0 +1,10 @@
+﻿using Dfe.Academies.Academisation.Core;
+using Dfe.Academies.Academisation.IService.ServiceModels.Legacy.ProjectAggregate;
+
+namespace Dfe.Academies.Academisation.IService.Commands.Legacy.Project
+{
+	public interface ILegacyProjectDeleteNoteCommand
+	{
+		Task<CommandResult> Execute(int projectId, ProjectNoteServiceModel note);
```
Comment (dneed-nimble) [inline Dfe.Academies.Academisation.IService/Commands/Legacy/Project/ILegacyProjectDeleteNoteCommand.cs:8]: Why is there two DeleteNote commands?

File: Dfe.Academies.Academisation.Data/ProjectAggregate/ProjectNoteAddCommand.cs (hunk 3 above)
Comment (tim-wilde-nimble) [reply Dfe.Academies.Academisation.Data/ProjectAggregate/ProjectNoteAddCommand.cs:18]: The `ProjectNote` type is the domain layer representation of the note. In the domain the relationship between a project and its notes is modelled as a collection of notes on the project, so a property for a foreign key is not required on the `ProjectNote`.

In the data layer the project note is represented with the `ProjectNoteState` type which - as it maps to rows in the db - _does_ have the ProjectId property as it maps to the foreign key column in the related table.

TL;DR - it's not in `ProjectNote` to ensure the database structure doesn't leak into the domain model.

File: Dfe.Academies.Academisation.IService/Commands/Legacy/Project/ILegacyProjectDeleteNoteCommand.cs (hunk 4 above)
Comment (tim-wilde-nimble) [reply Dfe.Academies.Academisation.IService/Commands/Legacy/Project/ILegacyProjectDeleteNoteCommand.cs:8]: This is the crux of the refactor, really. The refactor was because I had used the `AcademisationContext` outside of the data project/layer which is undesirable.

So; moved the command to the data layer (`ProjectNoteAddCommand` and `ProjectNoteDeleteCommand` where access to the `AcademisationContext` doesn't cross project boundaries) but the commands used at "higher" levels of the codebase are in the `Service` project. So - to keep the codebase consistent - there are these small wrapper commands here, too.

I suppose you could think of it like this: the ones in the `Data` project are the data commands, the ones in the service are the "domain action" commands and the controllers use domain action commands, not data commands directly.

File: Dfe.Academies.Academisation.Data/ProjectAggregate/ProjectNoteAddCommand.cs (hunk 3 above)
Comment (dneed-nimble) [reply Dfe.Academies.Academisation.Data/ProjectAggregate/ProjectNoteAddCommand.cs:18]: Perfect, makes sense.

File: Dfe.Academies.Academisation.IService/Commands/Legacy/Project/ILegacyProjectDeleteNoteCommand.cs (hunk 4 above)
Comment (dneed-nimble) [reply Dfe.Academies.Academisation.IService/Commands/Legacy/Project/ILegacyProjectDeleteNoteCommand.cs:8]: The logic makes sense but can't say I love the pattern!

Comment (tim-wilde-nimble) [reply Dfe.Academies.Academisation.IService/Commands/Legacy/Project/ILegacyProjectDeleteNoteCommand.cs:8]: Granted, in this case, the code is that "thin" that this looks odd, so is definitely more for wider codebase design consistency.

**Review by dneed-nimble — APPROVED** [review]
